websockify/websocketproxy.py

```diff
@@ -228,22 +233,43 @@ def do_proxy(self, target):
                                 self.server.target_host, self.server.target_port)
                     raise self.CClose(closed['code'], closed['reason'])
 
-
             if target in outs:
                 # Send queued client data to the target
                 dat = tqueue.pop(0)
-                sent = target.send(dat)
+                try:
+                    sent = target.send(dat)
+                except ssl.SSLWantWriteError:
+                    # nothing was sent - sending needs to be retried later
+                    sent = 0
+
                 if sent == len(dat):
                     self.print_traffic(">")
                 else:
                     # requeue the remaining data
                     tqueue.insert(0, dat[sent:])
                     self.print_traffic(".>")
 
-
             if target in ins:
                 # Receive target data, encode it and queue for client
-                buf = target.recv(self.buffer_size)
+                try:
+                    # It is strictly required that buffer size is more than 16kb, so that
+                    # all possible data can be received from a SSL socket in a single call.
+                    # Otherwise, there could be still data available for reading, but select
+                    # wouldn't report the socket as readable again, since all data has already
+                    # been read by the SSL socket from the OS.
+                    # see also: https://docs.python.org/3/library/ssl.html#notes-on-non-blocking-sockets
```

Comment:
I don't think it guarantees to give you everything in a single `recv()`. For one thing, there might be multiple frames buffered. So you need to do what that link specifies and continue calling `recv()` until it is drained. You have `SSLSocket.pending()` to help you out.

Reply:
Yes, it does. That's the good thing. Let me quote the documentation on OpenSSL's `SSL_read()`:

> The read functions work based on the SSL/TLS records. The data are received in records (with a maximum record size of 16kB). Only when a record has been completely received, can it be processed (decryption and check of integrity). Therefore, data that was not retrieved at the last read call can still be buffered inside the SSL layer and will be retrieved on the next read call. If num is higher than the number of bytes buffered then the read functions will return with the bytes buffered. If no more bytes are in the buffer, the read functions will trigger the processing of the next record. Only when the record has been received and processed completely will the read functions return reporting success. At most the contents of one record will be returned. 

And unfortunately, `SSLSocket.pending()` doesn't seem to work at all on some platforms. For me, it always returns zero regardless if there's something pending or not.

Reply:
The way I read that text it supports my initial assertion. If there are multiple records buffered then it states that only the first record will be returned. So you need to call it again.